### ISSUE TITLE
Application name defaults to "Saturn" majorly breaking framework expectations on this  

### ISSUE THREAD

**Issue by NinoFloris** [SaturnFramework/Saturn#69]
Things like IHostingStartup auto discovery and IStartup discovery fail, also distributed cache and quite a few other things expect this to be unique/equal to your startup assembly name. This stuff is for instance used by ApplicationInsights auto wireup.

The ApplicationName comes from the assembly that hosts the Configure delegate passed into `builder.Configure` as can be seen here https://github.com/aspnet/Hosting/blob/ae9da9290eb825f4145aaaa23603186d51211e53/src/Microsoft.AspNetCore.Hosting/WebHostBuilderExtensions.cs#L30

Temp fixes are:

- Change name inside `HostingEnvironment`
- Change it with `UseSetting(WebHostDefaults.ApplicationKey, "NewName")` before running build

PR fix would be to use the same reflection strategy `WebHostBuilderExtensions` uses on the router and override the setting before returning the builder.

**Comment by NinoFloris** [comment]
Grrr this won't work as it'll be compiled to a standard optimized closure, aka AssemblyName = FSharp.Core. 

```
    router (scope {
        forward "/api" apiScope
    })
```

Only like this will an actual type be generated inside your own app to get a good read of.
```
    router (fun next -> (next |> scope {
        forward "/api" apiScope
    }))
```

Any other clever ideas? @Krzysztof-Cieslak

**Comment by baronfel** [comment]
A possible workaround here would be for Saturn to add a `app_name` CE member to the `application` CE and set the backing setting explicitly, and if the user doesn't provide an app name then set it by default from the current executing assembly?

**Comment by NinoFloris** [comment]
Relevant, it may get more clear in 3.0 https://github.com/aspnet/AspNetCore/issues/7315

**Comment by davidalpert** [comment]
I'm confused why #14183 was closed as "no work to do" - can anyone shed light on that decision? This is the closest issue I can find.

I spent some time digging through the code in this repo trying to figure out why my ASPNETCORE 3.x app was not respecting the `APPLICATION_NAME` set up in my app configuration and learned that using the `DefaultWebHostBuilder` and configuring a `Startup` class injects an additional configuration provider (last so that it _overrides_ my app configuration) and forces the app name to the AssemblyName of the DLL containing the `Startup` class.

I'm curious why it does not simply respect the `APPLICATION_NAME` as configured by `IConfigurationBuilder` instance returned from the `ConfigureAppConfiguration` method?
